### PR TITLE
Update docs to reflect the fact that Iterable etc. are protocols

### LINKED ISSUE #4339: Update docs to reflect the fact that Iterable etc. are now protocols

### PR DESCRIPTION
**Pull request by JukkaL**
Also update the discussion of ABCs, and don't describe protocols as
experimental.

Fixes #4339.

### CODE REVIEW

File: docs/source/class_basics.rst [hunk 1]
```diff
@@ -267,6 +486,13 @@ be explicitly present if you are defining a protocol:
    # Error: nominal subtyping used by default
    x: NewProtocol = Concrete()  # Error!
 
+You can also include default implementations of methods in
+protocols. If you explicitly subclass these protocols you can inherit
+these default implementations. Explicitly including a protocol as a
+base class is also a way of documenting that your class implements a
+particular protocol, and it forces mypy to verify that your class
+implementation is actually compatible with the protocol.
```
Comment (ilevkivskyi) [inline docs/source/class_basics.rst]: There is a caveat that this still works (although it shouldn't according to PEP 544):
```python
class P(Protocol):
    def meth(self) -> int: ...  # No error here...

class C(P):
    pass

C()  # ...and no error here!
```
mypy currently treats empty body (ellipsis) as a default implementation. There is an issue to allow this only in stubs (and  in tests), but it is not implemented yet. (The current workaround is to use ``@abstractmethod``, IIRC the similar problem with literal ellipsis in ``x: int = ...`` is already fixed.)

File: docs/source/class_basics.rst [hunk 2]
```diff
@@ -174,16 +172,239 @@ a structural subtype of class ``C`` if the former has all attributes
 and methods of the latter, and with compatible types.
 
 Structural subtyping can be seen as a static equivalent of duck
-typing, which is well known to Python programmers. Mypy provides an
-opt-in support for structural subtyping via protocol classes described
+typing, which is well known to Python programmers. Mypy provides
+support for structural subtyping via protocol classes described
```
Comment (elazarg) [inline docs/source/class_basics.rst]: Nitpicking question: the part about duck typing is technically not true (e.g. cross relations between variables are not captured even when statically analysable). What does it give to the reader that is not conveyed by the rest of the discussion? If all we want is help readers that search for the term "duck typing", perhaps we can phrase is differently - maybe "Structural subtyping is a closer approximation of duck typing than inheritance" or some other weak claim.
(Sorry, I know this is not important; the rest of the PR LGTM).

File: docs/source/class_basics.rst (hunk 1 above)
Comment (ilevkivskyi) [reply docs/source/class_basics.rst]: For the reference here are the issues: https://github.com/python/mypy/issues/4066, https://github.com/python/mypy/issues/2350. You can increase priority if you think this is important.

Comment (gvanrossum) [reply docs/source/class_basics.rst]: I don't think we need to call this out in the docs. I do think it's important to fix. (Honestly I don't understand the difference between the two issues you link to.)

Comment (ilevkivskyi) [reply docs/source/class_basics.rst]: > (Honestly I don't understand the difference between the two issues you link to.)

They are duplicates, but both have some discussion, so I don't know which one to close.

File: docs/source/class_basics.rst [hunk 3]
```diff
@@ -174,16 +172,239 @@ a structural subtype of class ``C`` if the former has all attributes
...
+...............
+
+These protocols can be useful in async code.
+
+``Awaitable[T]``
+----------------
+
+.. code-block:: python
+
+   def __await__(self) -> Generator[Any, None, T]
```
[the remark this answers is not in the document]
Comment (ilevkivskyi) [reply docs/source/class_basics.rst]: I just know that some people (for example @1st1) don't like mentioning generators in async docs.

**Pull request merged**